In [191]:
import sqlite3 as sql
import pandas as pd

In [192]:
#connect to SQLite dataobase 
connection = sql.connect('twitch.db') #has 2 tables streamer data from 2021 and 2024
cursor = connection.cursor()

In [193]:
query = '''SELECT * FROM streamers2021;'''
df = pd.read_sql_query(query, connection)
df.head()

channel  watch_time  stream_time  peak_viewers  average_viewers  \
0     xQcOW  6196161750       215250        222720            27716   
1  summit1g  6091677300       211845        310998            25610   
2    Gaules  5644590915       515280        387315            10976   
3  ESL_CSGO  3970318140       517740        300575             7714   
4      Tfue  3671000070       123660        285644            29602   

   followers  followers_gained  views_gained partnered mature    language  \
0    3246298           1734810      93036735      True  False     English   
1    5310163           1370184      89705964      True  False     English   
2    1767635           1023779     102611607      True   True  Portuguese   
3    3944850            703986     106546942      True  False     English   
4    8938903           2068424      78998587      True  False     English   

  clean_name new_name normalized_name  
0      xqcow     None             xqc  
1   summit1g     None        summit1g  
2     gaules     None          gaules  
3    eslcsgo     None         eslcsgo  
4       tfue     None            tfue

The top 5 streamers watched streamers of 2021 are XQC, summit1g, Gaules, ESL_CSGO, and Tfue


The top 5 streamers with most followers in 2021 are Tfue, shroud, Myth, Rubius, and pokimane

In [194]:
#database columns
df.columns

Index(['channel', 'watch_time', 'stream_time', 'peak_viewers',
       'average_viewers', 'followers', 'followers_gained', 'views_gained',
       'partnered', 'mature', 'language', 'clean_name', 'new_name',
       'normalized_name'],
      dtype='object')

In [195]:
#streamers 2024
query = '''SELECT * FROM streamers2024;'''
df = pd.read_sql_query(query, connection)
df.head()

rank      name language         type              most_streamed_game  \
0     1  kaicenat  English  personality                   Just Chatting   
1     2    jynxzi  English  personality  Tom Clancy's Rainbow Six Siege   
2     3   caedrel  English  personality               League of Legends   
3     4   caseoh_  English  personality                        NBA 2K23   
4     5      ibai  Spanish  personality                   Just Chatting   

  second_most_streamed_game  average_stream_duration  \
0         I'm Only Sleeping                      7.6   
1                  NBA 2K20                      5.4   
2         I'm Only Sleeping                      6.3   
3             Just Chatting                      4.6   
4         League of Legends                      4.1   

   followers_gained_per_stream  avg_viewers_per_stream  avg_games_per_stream  \
0                        18405                   15852                   2.3   
1                         3386                    1145                   1.2   
2                          689                   12331                   1.3   
3                         7185                       0                   3.6   
4                         8289                  190714                   1.5   

   total_time_streamed  total_followers  total_views  total_games_streamed  \
0                 4698         10600000      9150000                   194   
1                 8407          5760000      1950000                    54   
2                 6728           797000     14200000                   111   
3                 2554          4220000           53                   385   
4                 6865         15600000    359000000                   149   

   active_days_per_week most_active_day day_with_most_followers_gained  \
0                   3.6          Friday                       Saturday   
1                   5.6         Tuesday                         Sunday   
2                   2.8        Thursday                         Sunday   
3                   6.2          Friday                         Monday   
4                   4.3       Wednesday                       Saturday   

  clean_name new_name normalized_name  
0   kaicenat     None        kaicenat  
1     jynxzi     None          jynxzi  
2    caedrel     None         caedrel  
3     caseoh     None          caseoh  
4       ibai     None            ibai

Top 5 ranking streamers are kaicenat, jynxzi, caedrel, caseoh_, and ibai

In [196]:
df.columns

Index(['rank', 'name', 'language', 'type', 'most_streamed_game',
       'second_most_streamed_game', 'average_stream_duration',
       'followers_gained_per_stream', 'avg_viewers_per_stream',
       'avg_games_per_stream', 'total_time_streamed', 'total_followers',
       'total_views', 'total_games_streamed', 'active_days_per_week',
       'most_active_day', 'day_with_most_followers_gained', 'clean_name',
       'new_name', 'normalized_name'],
      dtype='object')

In [197]:
#gained the most followers in 2021
query = '''SELECT channel, MAX(followers_gained)
FROM streamers2021
GROUP BY channel
ORDER BY MAX(followers_gained) DESC
LIMIT 5;'''
df = pd.read_sql_query(query, connection)
df

channel  MAX(followers_gained)
0  auronplay                3966525
1     Rubius                3820532
2   TheGrefg                3593081
3      Bugha                2220765
4   pokimane                2085831

In [198]:
#top 10 streams that got the highest viewship in 2021
query = '''SELECT channel, peak_viewers
FROM streamers2021
ORDER BY peak_viewers DESC
LIMIT 10;'''
df = pd.read_sql_query(query, connection)
df

channel  peak_viewers
0  Riot Games (riotgames)        639375
1                TheGrefg        538444
2                 dota2ti        483530
3                  shroud        471281
4              dota2ti_ru        457060
5                  Gaules        387315
6               csgomc_ru        364816
7                Warframe        346131
8        StarLadder_cs_en        329195
9                JokerdTV        319862

In [199]:
#2021 top streamers organized by follower count
query = '''SELECT channel, followers
from streamers2021
ORDER BY followers DESC
LIMIT 10;'''
df = pd.read_sql_query(query, connection)
df

channel  followers
0                    Tfue    8938903
1                  shroud    7744066
2                    Myth    6726893
3                  Rubius    5751354
4                pokimane    5367605
5                summit1g    5310163
6            TimTheTatman    5265659
7                 dakotaz    4520305
8  Riot Games (riotgames)    4487489
9            DrDisrespect    4450718

In [200]:
#top 25 streamers of 2024
query = '''SELECT name, total_followers
FROM streamers2024
LIMIT 25;'''
df = pd.read_sql_query(query, connection)
df

name  total_followers
0        kaicenat         10600000
1          jynxzi          5760000
2         caedrel           797000
3         caseoh_          4220000
4            ibai         15600000
..            ...              ...
20            xqc         12000000
21   loud_coringa          5380000
22  raderaderader           689000
23         gaules          4139999
24        yuuri22           564000

[25 rows x 2 columns]

In [201]:
#most categories in 2024
query = '''SELECT most_streamed_game
FROM streamers2024 
GROUP BY most_streamed_game
ORDER BY COUNT(*) DESC
LIMIT 10;'''
df = pd.read_sql_query(query, connection)
df

most_streamed_game
0       Just Chatting
1   League of Legends
2  Grand Theft Auto V
3            VALORANT
4              Casino
5            Fortnite
6              Dota 2
7      Counter-Strike
8           Minecraft
9      Virtual Casino

In [202]:
#Demographic of LOL streamers2024
query = '''SELECT language, COUNT(*) as count 
FROM streamers2024
WHERE most_streamed_game = 'League of Legends' 
GROUP BY language 
ORDER BY COUNT(*) DESC;'''
df = pd.read_sql_query(query, connection)
df

language  count
0      English     28
1   Portuguese     12
2       French     12
3      Spanish      9
4     Japanese      7
..         ...    ...
6       Polish      4
7       German      2
8      Turkish      1
9        Czech      1
10   Cantonese      1

[11 rows x 2 columns]

In [203]:
#Best day to gain more followers
query = '''SELECT day_with_most_followers_gained, COUNT(*) as count
FROM streamers2024
GROUP BY day_with_most_followers_gained 
ORDER BY COUNT(*) DESC;'''
df = pd.read_sql_query(query, connection)
df

day_with_most_followers_gained  count
0                         Sunday    358
1                       Saturday    159
2                         Monday    138
3                        Tuesday    127
4                         Friday     78
5                      Wednesday     74
6                       Thursday     65

In [204]:
#joins top 1000 streamers of 2021 with 2024 based on who is still in the top 1000
query = '''SELECT
s21.channel AS streamer,
s21.average_viewers AS avg_viewers_2021,
s24.avg_viewers_per_stream AS avg_viewers_2024,
s21.followers AS followers_2021,
s24.total_followers AS followers_2024
FROM streamers2021 s21
JOIN streamers2024 s24
ON LOWER(s21.channel) = LOWER(s24.name);'''
df = pd.read_sql_query(query, connection)
df


streamer  avg_viewers_2021  avg_viewers_2024  followers_2021  \
0        summit1g             25610            162001         5310163   
1          Gaules             10976            235565         1767635   
2       NICKMERCS             24181            114372         4074287   
3       loltyler1             22381            178397         3530767   
4           LIRIK             21739            117105         2666382   
..            ...               ...               ...             ...   
260     Staryuuki              1467             26495          586475   
261  LenaGol0vach              4159             18774          274237   
262      Surefour              2289              8004          331662   
263    aminematue              4016             30366          124232   
264     LITkillah              9104             69452          601927   

     followers_2024  
0           6270000  
1           4139999  
2           6720000  
3           5310000  
4           2970000  
..              ...  
260         3360000  
261          585000  
262          461000  
263         2590000  
264         2680000  

[265 rows x 5 columns]

In [205]:
#the number of streamers who kept their spot in top 1000 from 2021 to 2024
query = '''SELECT
CASE WHEN s21.channel IS NULL THEN 'New in 2024'
ELSE 'Returning from 2021'
END AS status,
COUNT(*) AS num_streamers
FROM streamers2024 s24
LEFT JOIN streamers2021 s21
ON LOWER(s24.name) = LOWER(s21.channel)
GROUP BY status;'''
df = pd.read_sql_query(query, connection)
df

status  num_streamers
0          New in 2024            734
1  Returning from 2021            265

At quick glance we see that there are only 265 streamers who were able to keep their spot in the top 1000 which shows how fast the streaming industry can change

In [206]:
#Top new streamers ordered by rank
query = '''SELECT
s24.rank,
s24.name,
s24.total_followers
FROM streamers2024 s24
LEFT JOIN streamers2021 s21
ON LOWER(s24.name) = LOWER(s21.channel)
WHERE s21.channel IS NULL
ORDER BY s24.rank
LIMIT 20;'''
df = pd.read_sql_query(query, connection)
df

rank            name  total_followers
0      1        kaicenat         10600000
1      2          jynxzi          5760000
2      3         caedrel           797000
3      4         caseoh_          4220000
4      8           tarik          3110000
..   ...             ...              ...
15    26        eliasn97          1880000
16    29       zackrawrr          1790000
17    31      ow_esports          1970000
18    33  legendus_shaka            29100
19    34  paulinholokobr          2930000

[20 rows x 3 columns]

Manually checking here we can see some familiar names however they are recorded differently than 2021 and this might be becuase they changed their channel username or recording differences.

In [207]:
#query for who is no longer in the top 1000
query = '''SELECT
s21.channel,
s21.followers
FROM streamers2021 s21
LEFT JOIN streamers2024 s24
ON LOWER(s21.channel) = LOWER(s24.name)
WHERE s24.name IS NULL
ORDER BY s21.followers DESC
limit 20;'''
df = pd.read_sql_query(query, connection)
df

channel  followers
0                     Tfue    8938903
1                     Myth    6726893
2             TimTheTatman    5265659
3                  dakotaz    4520305
4   Riot Games (riotgames)    4487489
..                     ...        ...
15             Castro_1021    2411995
16                 cloakzy    2138294
17             TSM_Hamlinz    2137521
18               CouRageJD    2115596
19               CDNThe3rd    2009972

[20 rows x 2 columns]

Here we will clean the dataset and match the names that are in both datasets but recorded differently so we can try to get an accurate look at who is actually still in the top 1000

In [208]:
query = '''SELECT
    LOWER(REPLACE(REPLACE(channel, '_', ''), ' ', '')) AS clean_name
FROM streamers2021;
'''
df = pd.read_sql_query(query, connection)
df

clean_name
0          10000days
1           1drakonz
2              1pvcs
3             39daph
4              72hrs
..               ...
995    핑맨(95pingman)
996  한동숙(handongsuk)
997   햇살살(hatsalsal)
998      헤징(hejin00)
999    홍방장(game2eye)

[1000 rows x 1 columns]

In [209]:
query = '''SELECT
    LOWER(REPLACE(REPLACE(name, '_', ''), ' ', '')) AS clean_name
FROM streamers2024;
'''
df = pd.read_sql_query(query, connection)
df

clean_name
0       kaicenat
1         jynxzi
2        caedrel
3         caseoh
4           ibai
..           ...
994        jlcs2
995      nobruxy
996      buxexat
997  sweetdreams
998  giantwaffle

[999 rows x 1 columns]

In [210]:
#Adding cleaned names to both databases
cursor.execute('''ALTER TABLE streamers2021
ADD COLUMN clean_name TEXT;''')

cursor.execute('''ALTER TABLE streamers2024
ADD COLUMN clean_name TEXT;''')

connection.commit()

cursor.execute('''UPDATE streamers2021
SET clean_name = LOWER(REPLACE(REPLACE(channel, '_', ''), ' ', ''));''')

cursor.execute('''UPDATE streamers2024
SET clean_name = LOWER(REPLACE(REPLACE(name, '_', ''), ' ', ''));''')

connection.commit()



OperationalError: duplicate column name: clean_name

Now we can see that the names for both datasets have bveen cleaned and normalized

In [ ]:
pd.read_sql("""
SELECT name, clean_name
FROM streamers2024
LIMIT 10
""", connection)


name  clean_name
0    kaicenat    kaicenat
1      jynxzi      jynxzi
2     caedrel     caedrel
3     caseoh_      caseoh
4        ibai        ibai
5   auronplay   auronplay
6     zerator     zerator
7       tarik       tarik
8   riotgames   riotgames
9  papaplatte  papaplatte

In [ ]:
pd.read_sql("""
SELECT channel, clean_name
FROM streamers2021
LIMIT 10
""", connection)

channel  clean_name
0       xQcOW       xqcow
1    summit1g    summit1g
2      Gaules      gaules
3    ESL_CSGO     eslcsgo
4        Tfue        tfue
5   Asmongold   asmongold
6   NICKMERCS   nickmercs
7  Fextralife  fextralife
8   loltyler1   loltyler1
9     Anomaly     anomaly

Now we can go in and fuzzy match the names to see account for people who changed their usernames or appear differently in each dataset

In [ ]:
#these are names that are similar to each other in both datasets
#now we manually go through them and see which names are actually the same and then map them to each other so they appear the same in both datasets
from itables import show

query = '''SELECT s2021.channel as channel2021, s2024.name as name2024, s2021.clean_name as clean2021, s2024.clean_name as clean2024
FROM streamers2021 s2021
JOIN streamers2024 s2024
ON (
       s2024.clean_name LIKE '%' || s2021.clean_name || '%'
    OR s2021.clean_name LIKE '%' || s2024.clean_name || '%'
   )
WHERE clean2021 <> clean2024;'''
df = pd.read_sql_query(query, connection)
show(df)



In [ ]:
#List of people that are the same but have different usernames or represented differently
# (xqcow, xqc
# riotgames(riotgames), riotgames
# gronkh, gronkhtv
# 蝦愛橘子(shuteyeorange), shuteyeorange
# 老皮(mobilmobil), mobilmobil
# 餐餐自由配(blusewillyretry), blusewillyretry
# skipnholive, skipnho
# 接接(godjj), godjj
# followgrubby, grubby
# 嬌兔(zrush), zrush
# 亞洲統神(asiagodtonegg3be0), asiagodtonegg3be0
# antoinedaniellive, antoinedaniel
# g4grevenant, revenant
# スタンミ(sutanmi), sutanmi
# 六希夫(qttsix), qttsix
# jlbichouu, bichouu
# 羅傑(roger9527), roger9527
# 超負荷(sam1268), sam1268)

Using this I can create a new table to standardize the names in both tables so they are represented in the same way

In [ ]:
cursor.execute('''drop TABLE if exists name_map;''') 

connection.commit()

In [ ]:
#create a table containing the mapped names
cursor.execute('''CREATE TABLE name_map (
               old_name TEXT PRIMARY KEY, 
               new_name TEXT);''') 

connection.commit()

cursor.execute('''INSERT INTO name_map (old_name, new_name) VALUES
               ('xqcow', 'xqc'),
               ('riotgames(riotgames)', 'riotgames'),
               ('gronkh', 'gronkhtv'),
               ('蝦愛橘子(shuteyeorange)', 'shuteyeorange'),
               ('老皮(mobilmobil)', 'mobilmobil'),
               ('餐餐自由配(blusewillyretry)', 'blusewillyretry'),
               ('skipnholive', 'skipnho'),
               ('接接(godjj)', 'godjj'),
               ('followgrubby', 'grubby'),
               ('嬌兔(zrush)', 'zrush'),
               ('亞洲統神(asiagodtonegg3be0)', 'asiagodtonegg3be0'),
               ('antoinedaniellive', 'antoinedaniel'),
               ('g4grevenant', 'revenant'),
               ('スタンミ(sutanmi)', 'sutanmi'),
               ('六希夫(qttsix)', 'qttsix'),
               ('jlbichouu', 'bichouu'),
               ('羅傑(roger9527)', 'roger9527'),
               ('超負荷(sam1268)', 'sam1268');''')

connection.commit()

#adds new column where the normalized names will go 
cursor.execute('''ALTER TABLE streamers2021 ADD COLUMN normalized_name TEXT;''')
cursor.execute('''ALTER TABLE streamers2024 ADD COLUMN normalized_name TEXT;''')

#adds the values to the column for both datasets
cursor.execute('''UPDATE streamers2021
               SET normalized_name = clean_name;''')

cursor.execute('''UPDATE streamers2024
               SET normalized_name = clean_name;''')

cursor.execute('''UPDATE streamers2021
               SET normalized_name = (
               SELECT new_name
               FROM name_map
               WHERE old_name = streamers2021.clean_name)
               WHERE clean_name IN (SELECT old_name FROM name_map);''')

cursor.execute('''UPDATE streamers2024
               SET normalized_name = (
               SELECT new_name
               FROM name_map
               WHERE old_name = streamers2024.clean_name)
               WHERE clean_name IN (SELECT old_name FROM name_map);''')

connection.commit()

Now the channel names that are supposed to be the same but appeared different in both tables should now be the same and we can get an accurate look at who is still in the top 1000. Counting the number of returning vs new top 1000 streamers we get

In [ ]:
query = '''SELECT
CASE WHEN s21.normalized_name IS NULL THEN 'New in 2024'
ELSE 'Returning from 2021'
END AS status,
COUNT(*) AS num_streamers
FROM streamers2024 s24
LEFT JOIN streamers2021 s21
ON s24.normalized_name = s21.normalized_name
GROUP BY status;'''
df = pd.read_sql_query(query, connection)
df

status  num_streamers
0          New in 2024            717
1  Returning from 2021            283

Below we can see all the people who kept their spot in the top 1000 alongside their cleaned names in each table

In [ ]:
query = '''SELECT DISTINCT
    s24.normalized_name,
    s24.clean_name AS name_2024,
    s21.clean_name AS name_2021
FROM streamers2024 s24
JOIN streamers2021 s21
ON TRIM(LOWER(s24.normalized_name)) = TRIM(LOWER(s21.normalized_name))
ORDER BY s24.normalized_name;
'''
df = pd.read_sql_query(query, connection)
show(df)
